# Основы глубинного обучения, майнор ИАД

## Домашнее задание 1. Введение в PyTorch. Полносвязные нейронные сети.

### Общая информация

Дата выдачи: 01.10.2023

Мягкий дедлайн: 23:59MSK 15.10.2023

Жесткий дедлайн: 23:59MSK 20.10.2023

### Оценивание и штрафы
Максимально допустимая оценка за работу — 10 баллов. За каждый день просрочки снимается 1 балл. Сдавать задание после жёсткого дедлайна сдачи нельзя.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов. Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

Неэффективная реализация кода может негативно отразиться на оценке.
Также оценка может быть снижена за плохо читаемый код и плохо оформленные графики. Все ответы должны сопровождаться кодом или комментариями о том, как они были получены.

Итогова оценка считается как
$$
min(task_1, task_2)*0.8 + max(task_1, task_2)*0.2
$$

где task_1 и task_2 - оценки за первое и второе заданиее соответсвенно.
Также, за домашнее задание выставляется 0, если не сделано нулевое или третье задание.
### О задании

В этом задании вам предстоит предсказывать год выпуска песни (**задача регрессии**) по некоторым звуковым признакам: [данные](https://archive.ics.uci.edu/ml/datasets/yearpredictionmsd). В ячейках ниже находится код для загрузки данных. Обратите внимание, что обучающая и тестовая выборки располагаются в одном файле, поэтому НЕ меняйте ячейку, в которой производится деление данных.

In [2]:
import torch
from torch import nn
import torch.nn.functional as F
import pandas as pd
import numpy as np
import random
device = "cuda" if torch.cuda.is_available() else "cpu"
from tqdm.notebook import tqdm
from IPython.display import clear_output
import matplotlib.pyplot as plt

In [3]:
!wget -O data.txt.zip https://archive.ics.uci.edu/ml/machine-learning-databases/00203/YearPredictionMSD.txt.zip

--2024-03-08 17:17:25--  https://archive.ics.uci.edu/ml/machine-learning-databases/00203/YearPredictionMSD.txt.zip
Распознаётся archive.ics.uci.edu (archive.ics.uci.edu)… 128.195.10.252
Подключение к archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... соединение установлено.
HTTP-запрос отправлен. Ожидание ответа… 200 OK
Длина: нет данных
Сохранение в: «data.txt.zip»

data.txt.zip            [             <=>    ] 201,24M  16,1MB/s    за 17s     

2024-03-08 17:17:43 (11,7 MB/s) - «data.txt.zip» сохранён [211011981]



In [4]:
df = pd.read_csv('data.txt.zip', header=None)
df.head(10)

0         1         2         3         4         5         6         7   \
0  2001  49.94357  21.47114  73.07750   8.74861 -17.40628 -13.09905 -25.01202   
1  2001  48.73215  18.42930  70.32679  12.94636 -10.32437 -24.83777   8.76630   
2  2001  50.95714  31.85602  55.81851  13.41693  -6.57898 -18.54940  -3.27872   
3  2001  48.24750  -1.89837  36.29772   2.58776   0.97170 -26.21683   5.05097   
4  2001  50.97020  42.20998  67.09964   8.46791 -15.85279 -16.81409 -12.48207   
5  2001  50.54767   0.31568  92.35066  22.38696 -25.51870 -19.04928  20.67345   
6  2001  50.57546  33.17843  50.53517  11.55217 -27.24764  -8.78206 -12.04282   
7  2001  48.26892   8.97526  75.23158  24.04945 -16.02105 -14.09491   8.11871   
8  2001  49.75468  33.99581  56.73846   2.89581  -2.92429 -26.44413   1.71392   
9  2007  45.17809  46.34234 -40.65357  -2.47909   1.21253  -0.65302  -6.95536   

         8         9   ...        81         82         83        84  \
0 -12.23257   7.83089  ...  13.01620  -54.40548   58.99367  15.37344   
1  -0.92019  18.76548  ...   5.66812  -19.68073   33.04964  42.87836   
2  -2.35035  16.07017  ...   3.03800   26.05866  -50.92779  10.93792   
3 -10.34124   3.55005  ...  34.57337 -171.70734  -16.96705 -46.67617   
4  -9.37636  12.63699  ...   9.92661  -55.95724   64.92712 -17.72522   
5  -5.19943   3.63566  ...   6.59753  -50.69577   26.02574  18.94430   
6  -9.53930  28.61811  ...  11.63681   25.44182  134.62382  21.51982   
7  -1.87566   7.46701  ...  18.03989  -58.46192  -65.56438  46.99856   
8  -0.55644  22.08594  ...  18.70812    5.20391  -27.75192  17.22100   
9 -12.20040  17.02512  ...  -4.36742  -87.55285  -70.79677  76.57355   

         85        86         87        88         89        90  
0   1.11144 -23.08793   68.40795  -1.82223  -27.46348   2.26327  
1  -9.90378 -32.22788   70.49388  12.04941   58.43453  26.92061  
2  -0.07568  43.20130 -115.00698  -0.05859   39.67068  -0.66345  
3 -12.51516  82.58061  -72.08993   9.90558  199.62971  18.85382  
4  -1.49237  -7.50035   51.76631   7.88713   55.66926  28.74903  
5  -0.33730   6.09352   35.18381   5.00283  -11.02257   0.02263  
6   8.17570  35.46251   11.57736   4.50056   -4.62739   1.40192  
7  -4.09602  56.37650  -18.29975  -0.30633    3.98364  -3.72556  
8  -0.85210 -15.67150  -26.36257   5.48708   -9.13495   6.08680  
9  -7.71727   3.26926 -298.49845  11.49326  -89.21804 -15.09719  

[10 rows x 91 columns]

Мы вывели кусок данных, чтобы понять, насколько они пригодны для работы без изменений. Здесь ясно, что сомнительно дальше с такими данными работать, потому что как минимум есть отрицательные значения, которые не отмасштабированы, кроме того еще сразу бросается в глаза совсем разная размерность, где-то видим реально большие числа, а где-то 0.075. Ясно, что будем скейлить.

In [5]:
X = df.iloc[:, 1:].values
y = df.iloc[:, 0].values

train_size = 463715
X_train = X[:train_size, :]
y_train = y[:train_size]
X_test = X[train_size:, :]
y_test = y[train_size:]

## Задание 0. (0 баллов, но при невыполнении максимум за все задание &mdash; 0 баллов)

Мы будем использовать RMSE как метрику качества. Для самого первого бейзлайна обучите `Ridge` регрессию из `sklearn`. Кроме того, посчитайте качество при наилучшем константном прогнозе.

Для выполнения данного задания (и всех последующих) предобработайте данные.

1. Зафиксируйте random_seed везде где только возможно. Вам предоставлена функция для этого, однако вы можете дополнить ее своими дополнениями
2. Обучите StandertScaler и предобработайте ваши данные. В следующих заданиях можете использовать другой scaler или вообще отказаться от него


### Scaling 

In [6]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()


In [7]:
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)


### Selecting random seed

In [8]:
def set_random_seed(seed):
    torch.backends.cudnn.deterministic = True
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    np.random.seed(seed)
    random.seed(seed)
    # Add smth here if you need...

In [9]:
set_random_seed(505505)

### Modeling 

In [10]:
from sklearn.linear_model import RidgeCV

In [11]:
model = RidgeCV(alphas = np.array([100, 1000]), fit_intercept= True, cv = 5)

In [12]:
model.fit(X_train, y_train)

RidgeCV(alphas=array([ 100, 1000]), cv=5)

In [13]:
y_pred = model.predict(X_test)

In [14]:
from sklearn.metrics import mean_squared_error

In [15]:
rmse = pow( mean_squared_error(y_pred, y_test), 0.5)

In [16]:
rmse

9.510172609963114

Лучшая константа для RMSE это среднее, посчитаем значение метрики при нем

In [17]:
y_mean = np.full(X_test.shape[0], y_train.mean())

In [18]:
best_rmse_metric = pow( mean_squared_error(y_mean, y_test), 0.5)

In [19]:
best_rmse_metric

10.85246390513634

## Задание 1. (максимум 10 баллов)

Закрепите свои знания о том, как pytorch работает с обратным распространением ошибки, проделав следующие шаги:

1. Создайте модель линейной регрессии, которая будет состоять только из одного Linear слоя.
2. Напишите цикл обучения вашей линейной регрессии. В нем реализуйте подсчет функции потерь, сделайте шаг градиентного спуска. Запрещено использовать готовые оптимизаторы и loss-функции из библиотеки pytorch. Для подсчета градиента воспользуйтесь методом backward.
3. Запустите обучение на 10 эпохах, после каждой проверяйте значение целевой метрики на тестовой выборке.
4. Выведите на экран графики метрики и значения функции потерь на тестовой и обучающей выборке.

В данном задании нет цели побить какой-то порог по метрике. Ваша задача - убедиться в том, что ваш рукописный цикл обучения работает. Для ускорения вычислений и обучения модели можете брать только срез данных, а не весь датасет.

### Data

In [20]:
class myDataset(torch.utils.data.Dataset):
    
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __len__(self):
        return self.x.shape[0]
    
    def __getitem__(self, idx):
        return torch.tensor(self.x[idx, :], dtype=torch.float), self.y[idx]

In [21]:
X_tr = torch.tensor(X_train)
X_tst = torch.tensor(X_test)
Y_tr = torch.tensor(y_train)
Y_tst = torch.tensor(y_test)


In [22]:
data = myDataset(X_tr, Y_tr)

In [23]:
data_test = myDataset(X_test, Y_tst)

In [24]:
dataloader = torch.utils.data.DataLoader(data, batch_size=10000)

In [25]:
dataloader_test = torch.utils.data.DataLoader(data_test, batch_size=len(data_test))

In [26]:
X_tr.shape

torch.Size([463715, 90])

In [27]:
Y_tr.shape

torch.Size([463715])

### Model

In [28]:
model = nn.Sequential(
    nn.Linear(
        X_tr.shape[1], 200
    ),
    nn.ReLU(),
    nn.Linear(X_tr.shape[1], 100), 
    nn.ReLU(),
    nn.Linear(X_tr.shape[1], 1), 

)

In [29]:
optimizer = torch.optim.SGD(model.parameters(), lr=0.005)

### Training

In [31]:
import wandb

In [33]:
wandb.init(project="pytorch-demo")
# сохраняем параметры сетки в wandb + просим следить за градиентами сетки
wandb.watch(model);

Failed to detect the name of this notebook, you can set it manually with the WANDB_NOTEBOOK_NAME environment variable to enable code saving.
wandb: Currently logged in as: to505to505 (to505to505_team). Use `wandb login --relogin` to force relogin


In [88]:
if torch.cuda.is_available():
    device = torch.device('cuda')
else:
    device = torch.device('cpu')

In [89]:
device

device(type='cpu')

In [39]:
counter = 0
for epoch in range(20):
    for x_train, y_train in tqdm(dataloader):
        counter+=1
        y_pred = model(x_train)
        criterion = torch.nn.MSELoss()
        loss = criterion(y_pred, y_train.float())
        loss.backward()
        optimizer.step()  
        optimizer.zero_grad()  
        val_loss = []  # сюда будем складывать **средний по бачу** лосс
        if(counter%10==0):
            with torch.no_grad():  # на валидации запрещаем фреймворку считать градиенты по параметрам
                for x_val, y_val in tqdm(dataloader_test):
                    y_pred = model(x_val.clone().detach().float())  
                    loss = criterion(y_pred, y_val.view(y_pred.shape))  
                    val_loss.append(loss.item())  # добавляем значение потери в список

            # скидываем метрики на wandb и автоматом смотрим на графики
            wandb.log(
                {
                    "mean val loss": np.mean(val_loss),
                }
            )

            # печатаем метрики
            print(
                f"Epoch: {epoch}, loss: {np.mean(val_loss)}"
            )

  0%|          | 0/47 [00:00<?, ?it/s]

/var/folders/wn/vk7vt0j13lgf017ktpdypywr0000gn/T/ipykernel_35282/1995826793.py:11: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  return torch.tensor(self.x[idx, :], dtype=torch.float), self.y[idx]


ValueError: not enough values to unpack (expected 3, got 2)

In [37]:
val_loss[-1]

117.62213897705078

In [38]:
torch.save(model.state_dict(), '3layer_first_try')

wandb: Network error (ConnectionError), entering retry loop.


## Задание 2. (максимум 10 баллов)

Реализуйте обучение и тестирование нейронной сети для предоставленного вам набора данных. Соотношение между полученным значением метрики на тестовой выборке и баллами за задание следующее:

- $\text{RMSE} \le 9.00 $ &mdash; 4 балла
- $\text{RMSE} \le 8.90 $ &mdash; 6 баллов
- $\text{RMSE} \le 8.80 $ &mdash; 8 баллов
- $\text{RMSE} \le 8.75 $ &mdash; 10 баллов

Есть несколько правил, которых вам нужно придерживаться:

- Весь пайплайн обучения должен быть написан на PyTorch. При этом вы можете пользоваться другими библиотеками (`numpy`, `sklearn` и пр.), но только для обработки данных. То есть как угодно трансформировать данные и считать метрики с помощью этих библиотек можно, а импортировать модели из `sklearn` и выбивать с их помощью требуемое качество &mdash; нельзя. Также нельзя пользоваться библиотеками, для которых сам PyTorch является зависимостью.

- Мы никак не ограничиваем ваш выбор архитектуры модели, но скорее всего вам будет достаточно полносвязной нейронной сети.

- Для обучения запрещается использовать какие-либо иные данные, кроме обучающей выборки.

- Ансамблирование моделей запрещено.

### Полезные советы:

- Очень вряд ли, что у вас с первого раза получится выбить качество на 10 баллов, поэтому пробуйте разные архитектуры, оптимизаторы и значения гиперпараметров. В идеале при запуске каждого нового эксперимента вы должны менять что-то одно, чтобы точно знать, как этот фактор влияет на качество.

- Не забудьте, что для улучшения качества модели вам поможет **нормировка таргета**.

- Тот факт, что мы занимаемся глубинным обучением, не означает, что стоит забывать про приемы, использующиеся в классическом машинном обучении. Так что обязательно проводите исследовательский анализ данных, отрисовывайте нужные графики и не забывайте про масштабирование и подбор гиперпараметров.

- Вы наверняка столкнетесь с тем, что ваша нейронная сеть будет сильно переобучаться. Для нейросетей существуют специальные методы регуляризации, например, dropout ([статья](https://jmlr.org/papers/volume15/srivastava14a/srivastava14a.pdf)) и weight decay ([блогпост](https://towardsdatascience.com/weight-decay-l2-regularization-90a9e17713cd)). Они, разумеется, реализованы в PyTorch. Попробуйте поэкспериментировать с ними.

- Если вы чего-то не знаете, не гнушайтесь гуглить. В интернете очень много полезной информации, туториалов и советов по глубинному обучению в целом и по PyTorch в частности. Но не забывайте, что за скатанный код без ссылки на источник придется ответить по всей строгости!

- Если вы сразу реализуете обучение на GPU, то у вас будет больше времени на эксперименты, так как любые вычисления будут работать быстрее. Google Colab предоставляет несколько GPU-часов (обычно около 8-10) в сутки бесплатно.

- Чтобы отладить код, можете обучаться на небольшой части данных или даже на одном батче. Если лосс на обучающей выборке не падает, то что-то точно идет не так!

- Пользуйтесь утилитами, которые вам предоставляет PyTorch (например, Dataset и Dataloader). Их специально разработали для упрощения разработки пайплайна обучения.

- Скорее всего вы захотите отслеживать прогресс обучения. Для создания прогресс-баров есть удобная библиотека `tqdm`.

- Быть может, вы захотите, чтобы графики рисовались прямо во время обучения. Можете воспользоваться функцией [clear_output](http://ipython.org/ipython-doc/dev/api/generated/IPython.display.html#IPython.display.clear_output), чтобы удалять старый график и рисовать новый на его месте.

**ОБЯЗАТЕЛЬНО** рисуйте графики зависимости лосса/метрики на обучающей и тестовой выборках в зависимости от времени обучения. Если обучение занимает относительно небольшое число эпох, то лучше рисовать зависимость от номера шага обучения, если же эпох больше, то рисуйте зависимость по эпохам. Если проверяющий не увидит такого графика для вашей лучшей модели, то он в праве снизить баллы за задание.

**ВАЖНО!** Ваше решение должно быть воспроизводимым. Если это не так, то проверяющий имеет право снизить баллы за задание. Чтобы зафиксировать random seed, воспользуйтесь функцией из предыдущего задания.



Вы можете придерживаться любой адекватной струкуры кода, но мы советуем воспользоваться сигнатурами функций, которые приведены ниже. Лучше всего, если вы проверите ваши предсказания ассертом: так вы убережете себя от разных косяков, например, что вектор предсказаний состоит из всего одного числа. В любом случае, внимательно следите за тем, для каких тензоров вы считаете метрику RMSE. При случайном или намеренном введении в заблуждение проверяющие очень сильно разозлятся.

In [ ]:
train_set = None  # YOUR CODE HERE (－.－)...zzzZZZzzzZZZ
train_loader = None  # YOUR CODE HERE (－.－)...zzzZZZzzzZZZ

test_set = None  # YOUR CODE HERE (－.－)...zzzZZZzzzZZZ
test_loader = None  # YOUR CODE HERE (－.－)...zzzZZZzzzZZZ

model = None  # YOUR CODE HERE (－.－)...zzzZZZzzzZZZ
optimizer = None  # YOUR CODE HERE (－.－)...zzzZZZzzzZZZ
criterion = None  # YOUR CODE HERE (－.－)...zzzZZZzzzZZZ

In [ ]:
def train(model, optimizer, criterion, train_loader, test_loader):
    """
    params:
        model - torch.nn.Module to be fitted
        optimizer - model optimizer
        criterion - loss function from torch.nn
        train_loader - torch.utils.data.Dataloader with train set
        test_loader - torch.utils.data.Dataloader with test set
                      (if you wish to validate during training)
    """
    # YOUR CODE HERE (－.－)...zzzZZZzzzZZZ
    raise NotImplementedError


def test(model, criterion, test_loader):
    """
    params:
        model - torch.nn.Module to be evaluated on test set
        criterion - loss function from torch.nn
        test_loader - torch.utils.data.Dataloader with test set
    ----------
    returns:
        predicts - torch.tensor with shape (len(test_loader.dataset), ),
                   which contains predictions for test objects
    """
    # YOUR CODE HERE (－.－)...zzzZZZzzzZZZ
    predicts = torch.ones(len(test_loader.dataset))
    return predicts


In [ ]:
assert test(model, criterion, test_loader).shape[0] == y_test.shape[0]

In [ ]:
# ೭੧(❛〜❛✿)੭೨ Go forward!

## Задание 3. (0 баллов, но при невыполнении максимум за все задание &mdash; 0 баллов)

Напишите небольшой отчет о том, как вы добились полученного качества: какие средства использовали и какие эксперименты проводили. Подробно расскажите об архитектурах и значениях гиперпараметров, а также какие метрики на тесте они показывали. Чтобы отчет был зачтен, необходимо привести хотя бы 3 эксперимента.